components/style/animation.rs

```diff
@@ -369,56 +365,182 @@ impl PropertyAnimation {
-    // if the property no longer transitions.
-    for running_animation in previously_running_animations.into_iter() {
-        if let Animation::Transition(_, _, ref property_animation) = running_animation {
-            if !properties_that_transition.contains(property_animation.property_id()) {
-                cancelled_animations.push(running_animation);
-                continue;
+#[derive(Default)]
+pub struct ElementAnimationState {
+    /// The animations running for this element.
+    pub running_animations: Vec<Animation>,
+
+    /// The animations that have finished for this element, but not canceled. These are cleared
+    /// upon triggering a DOM event for each finished animation.
+    pub finished_animations: Vec<Animation>,
+
+    /// The animations that have been cancelled for this element. These are cleared
+    /// upon triggering a DOM event for each cancelled animation.
+    pub cancelled_animations: Vec<Animation>,
+
+    /// New animations created for this element.
+    pub new_animations: Vec<Animation>,
+}
+
+#[cfg(feature = "servo")]
+impl ElementAnimationState {
+    /// Cancel all animations in this `ElementAnimationState`. This is typically called
+    /// when the element has been removed from the DOM.
+    pub fn cancel_all_animations(&mut self) {
+        self.cancelled_animations
+            .extend(self.finished_animations.drain(0..));
+        self.cancelled_animations
+            .extend(self.running_animations.drain(0..));
+        self.cancelled_animations
+            .extend(self.new_animations.drain(0..));
+    }
+
+    /// Whether this `ElementAnimationState` is empty, which means it doesn't
+    /// hold any animations in any state.
+    pub fn is_empty(&self) -> bool {
+        self.running_animations.is_empty() &&
+            self.finished_animations.is_empty() &&
+            self.cancelled_animations.is_empty() &&
+            self.new_animations.is_empty()
+    }
+
+    fn add_new_animation(&mut self, animation: Animation) {
+        self.new_animations.push(animation);
+    }
+
+    fn add_or_update_new_animation(&mut self, timer: &Timer, new_animation: Animation) {
+        // If the animation was already present in the list for the node,
+        // just update its state.
+        if let Animation::Keyframes(_, _, ref new_name, ref new_state) = new_animation {
+            for existing_animation in self.running_animations.iter_mut() {
+                match existing_animation {
+                    Animation::Keyframes(_, _, ref name, ref mut state) if *name == *new_name => {
+                        state.update_from_other(&new_state, timer);
+                        return;
+                    },
+                    _ => {},
+                }
             }
         }
-        running_animations.push(running_animation);
+        // Otherwise just add the new running animation.
+        self.add_new_animation(new_animation);
+    }
+}
+
+#[cfg(feature = "servo")]
```

Comment:
I don't get the usefulness of this as it exists, on this PR at least. This takes at least the same space as an `ElementAnimationState`, but creating an empty `ElementAnimationState` is cheap and doesn't allocate.

Am I missing something? Or is this just a footgun-preventor so that people can accidentally store empty long-lived `ElementAnimationState` objects?

Reply:
I agree and after our conversation on Matrix I've removed this data structure. Now there is only `ElementAnimationState`.